Add moving_fn pipeline aggregation (#9796)
opensearch-project/documentation-website: _aggregations/pipeline/moving_fn.md

| `format`              | Optional          | String          | A [DecimalFormat](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/text/DecimalFormat.html) formatting string. Returns the formatted output in the aggregation's `value_as_string` property. |
| `shift`               | Optional          | Integer         | The number of buckets to shift the window. Can be positive (shift right toward future buckles) or negative (toward past buckets). Default is `0`, which places the window immediately to the left of the current bucket. See [Shifting the window](#shifting-the-window). |

## How moving function works

The `moving_fn` aggregation operates on a sliding window over an ordered sequence of buckets. Starting at the first bucket in the parent aggregation, `moving_fn` does the following:

1. Collects the subsequence (window) of values from the buckets specified by the `window` and `shift` parameters.
2. Passes these values as an array to the function specified by `script`.
3. Uses `script` to compute a single value from the array.
4. Returns this value as the result for the current bucket.
5. Moves forward one bucket and repeats this process.

> "Past" and "future" values imply time-series data, the most common use case for moving window functions. More generally, you can take them to refer to previous and upcoming values, respectively, in any ordered data sequence.

Reviewer: ```suggestion
"Past" and "future" values imply time-series data, the most common use case for moving window functions. More generally, they refer to previous and upcoming values, respectively, in any ordered data sequence.
```

{: .note}

The script applied by `moving_fn` can be a [predefined function](#predefined-functions) or a [custom script](#custom-scripts). Bucket values are provided to the script in the `values` array. The script returns a double value as the result. The result values `NaN` and `+/- Inf` are allowed, but `null` is not.

### Window size

> The `window` parmeter specifies the number of buckets defining the size of the window.

Reviewer: ```suggestion
The `window` parameter specifies the number of buckets that define the size of the window.
```

The array passed to the `script` function is zero-indexed. Its values are accessed within the script as `values[0]` to `values[n]`, where `n = values.length - 1`.

### Shifting the window

> The `shift` parameter controls where the moving window lies relative to the current bucket. Set `shift` based on whether your analysis requires historical context, current data, or future prediction. The default is `0`, which shows only past values (excluding the current bucket). 

Reviewer: ```suggestion
The `shift` parameter controls where the moving window is located relative to the current bucket. Set `shift` based on whether your analysis requires historical context, current data, or future prediction. The default is `0`, which shows only past values (excluding the current bucket). 
```

Some commonly used values of `shift` are as follows:

| `shift` | Window description                            |                     |
| :--           | :--                                           | :--                 |
| `0`           | Only past values. Excludes the current value. | `--[-----]x----`    |
| `1`           | Past values, including the current value.     | `--[----x]-----`    |
| `window/2`    | Centers the window around the current value.  | `--[--x--]-----`    |
| `window`      | Future values, including the current value.   | `--[x----]-----`    |

When a window extends beyond available data at the beginning or end of a sequence, `window` shrinks automatically to use only the available points:

```
[x----]--
-[x----]-
--[x----]
---[x---]
----[x--]
-----[x-]
------[x]
```

## Predefined functions

> The `moving_fn` aggregation supports a number of predefined functions that can be used instead of a custom script. The functions are accessible from the `MovingFunctions` context. For example, access the `max` function as `MovingFunctions.max(values)`. 

Reviewer: ```suggestion
The `moving_fn` aggregation supports a number of predefined functions that can be used instead of a custom script. The functions are accessible from the `MovingFunctions` context. For example, you can access the `max` function as `MovingFunctions.max(values)`. 
```

| Linear weighted average               | `linearWeightedAvg`  | A weighted average using a linear decay of weights, giving more importance to recent values. |
| Exponentially Weighted Moving Average | `ewma`               | A weighted average using exponentially decaying weights, giving more importance to recent values. |
| Holt                                  | `holt`               | A weighted average using a second exponential term to smooth long-term trends. |
| Holt-Winters                          | `holt_wimnters`      | A weighted average using a third exponential term to smooth periodic (seasonal) effects. |
| Standard deviation                    | `stdDev`             | The sum of values in the window. |

All of the predefined functions take the `values` array as their first parameter. For functions that take extra parameters, pass these parameters in order after `values`. For example, call the `stdDev` function by setting the `script` value to `MovingFunctions.stdDev(values, MovingFunctions.unweightedAvg(values))`.

The following table shows the settings required for each model.

> | Function            | Extra parameters   | Allowed values  | Default | Description |

Reviewer 2: I think merging the cells (in `Function` column) with same values looks more clean. Check it out:

<table>
  <thead>
    <tr>
      <th>Function</th>
      <th>Extra parameters</th>
      <th>Allowed values</th>
      <th>Default</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>max</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>min</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>sum</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>unweightedAvg</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>linearWeightedAvg</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>ewma</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td rowspan="2" style="vertical-align: middle;"><code>holt</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td><code>beta</code></td>
      <td>[0, 1]</td>
      <td>0.1</td>
      <td>The trend decay parameter.</td>
    </tr>
    <tr>
      <td rowspan="6" style="vertical-align: middle;"><code>holt_winters</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td><code>beta</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The trend decay parameter.</td>
    </tr>
    <tr>
      <td><code>gamma</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The periodic decay parameter.</td>
    </tr>
    <tr>
      <td><code>type</code></td>
      <td><code>add</code>, <code>mult</code></td>
      <td><code>add</code></td>
      <td>How seasonality is incorporated.</td>
    </tr>
    <tr>
      <td><code>period</code></td>
      <td>Integer</td>
      <td>1</td>
      <td>The number of buckets comprising the period.</td>
    </tr>
    <tr>
      <td><code>pad</code></td>
      <td>Boolean</td>
      <td>true</td>
      <td>Whether to add a small offset to <code>0</code> values for <code>mult</code> type models to avoid a divide-by-zero error.</td>
    </tr>
    <tr>
      <td><code>stdDev</code></td>
      <td><code>avg</code></td>
      <td>Any double</td>
      <td>None</td>
      <td>Any double value, but to compute a meaningful standard deviation, use the mean of the sliding window array. Normally, <code>MovingFunctions.unweightedAvg(values)</code>.</td>
    </tr>
  </tbody>
</table>

I used a GPT tool to reformat this in html and combine/merge cells. Here is the code snippet which you can toggle or use directly:

```
<table>
  <thead>
    <tr>
      <th>Function</th>
      <th>Extra parameters</th>
      <th>Allowed values</th>
      <th>Default</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>max</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>min</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>sum</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>unweightedAvg</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>linearWeightedAvg</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>ewma</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td rowspan="2" style="vertical-align: middle;"><code>holt</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td><code>beta</code></td>
      <td>[0, 1]</td>
      <td>0.1</td>
      <td>The trend decay parameter.</td>
    </tr>
    <tr>
      <td rowspan="6" style="vertical-align: middle;"><code>holt_winters</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td><code>beta</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The trend decay parameter.</td>
    </tr>
    <tr>
      <td><code>gamma</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The periodic decay parameter.</td>
    </tr>
    <tr>
      <td><code>type</code></td>
      <td><code>add</code>, <code>mult</code></td>
      <td><code>add</code></td>
      <td>How seasonality is incorporated.</td>
    </tr>
    <tr>
      <td><code>period</code></td>
      <td>Integer</td>
      <td>1</td>
      <td>The number of buckets comprising the period.</td>
    </tr>
    <tr>
      <td><code>pad</code></td>
      <td>Boolean</td>
      <td>true</td>
      <td>Whether to add a small offset to <code>0</code> values for <code>mult</code> type models to avoid a divide-by-zero error.</td>
    </tr>
    <tr>
      <td><code>stdDev</code></td>
      <td><code>avg</code></td>
      <td>Any double</td>
      <td>None</td>
      <td>Any double value, but to compute a meaningful standard deviation, use the mean of the sliding window array. Normally, <code>MovingFunctions.unweightedAvg(values)</code>.</td>
    </tr>
  </tbody>
</table>
```

Reviewer 3: @sandeshkr419 Thanks for the suggestion. While I do agree that merged cells look better, we don't use HTML tables in the documentation unless absolutely necessary. Using HTML makes the table harder to edit. Additionally, our theme and supported css does not render merged cells correctly: the first row after a merged cell has a left border while the others do not. Since our upstream theme also defines table styles, we would have to override those by using `!important` and define the borders for all cell types, which, in my opinion, does not warrant using HTML. Here's a picture of how this renders in our current styling:
![Screenshot 2025-05-20 at 5 10 21 PM](https://github.com/user-attachments/assets/87789d84-1748-432b-9876-1794a4b1fa5e)

Reviewer 3: @sandeshkr419 I kept the markdown table but removed the duplicate values from the first column and augmented the descriptions so there are no blank cells. Please take a look.

| :--                 | :--                | :--             | :--     | :--         |
| `max`               | None               |                 |         |             |
| `min`               | None               |                 |         |             |
| `sum`               | None               |                 |         |             |
| `unweightedAvg`     | None               |                 |         |             |
| `linearWeightedAvg` | None               |                 |         |             |
| `ewma`              | `alpha`            | [0, 1] | 0.3     | The decay parameter. |
| `holt`              | `alpha`            | [0, 1] | 0.3     | The decay parameter. |
| `holt`              | `beta`             | [0, 1] | 0.1     | The trend decay parameter. |
| `holt_winters`      | `alpha`            | [0, 1] | 0.3     | The decay parameter. |
| `holt_winters`      | `beta`             | [0, 1] | 0.3     | The trend decay parameter. |
| `holt_winters`      | `gamma`            | [0, 1] | 0.3     | The periodic decay parameter. |
| `holt_winters`      | `type`             | `add`, `mult`   | `add`   | How seasonality is incorporated. |

> | `holt_winters`      | `period`           | Integer         | 1       | The number of buckets that comprise the period. |

Reviewer: ```suggestion
| `holt_winters`      | `period`           | Integer         | 1       | The number of buckets comprising the period. |
```

| `holt_winters`      | `pad`              | Boolean         | true    | Whether to add a small offset to `0` values for `mult` type models to avoid a divide-by-zero error. |

> | `stdDev`            | `avg`              | Any Double      | None    | Any Double value, but to compute a meaningful standard deviation use the mean of the sliding window array. Normally, `MovingFunctions.unweightedAvg(values)`. |

Reviewer: ```suggestion
| `stdDev`            | `avg`              | Any double      | None    | Any double value, but to compute a meaningful standard deviation, use the mean of the sliding window array. Normally, `MovingFunctions.unweightedAvg(values)`. |
```

            "value": 9544718
          },
          "the_movavg": {
            "value": 195603.33146038183
          }
        },
        {
          "key_as_string": "2025-05-12T00:00:00.000Z",
          "key": 1747008000000,
          "doc_count": 1610,
          "the_sum": {
            "value": 9155820
          },
          "the_movavg": {
            "value": 270085.92336040025
          }
        },
        {
          "key_as_string": "2025-05-19T00:00:00.000Z",
          "key": 1747612800000,
          "doc_count": 1610,
          "the_sum": {
            "value": 9025078
          },
          "the_movavg": {
            "value": 269477.75659701484
          }
        },
        {
          "key_as_string": "2025-05-26T00:00:00.000Z",
          "key": 1748217600000,
          "doc_count": 895,
          "the_sum": {
            "value": 5047345
          },
          "the_movavg": {
            "value": 267356.5422566652
          }
        }
      ]
    }
  }
}
```
</details>

## Custom scripts

You can supply an arbitrary custom script to calculate `moving_fn` results. Custom scripts use the Painless scripting language.

### Example: Custom scripts

> The following example creates a date histogram with a one-week interval from the OpenSearch Dashboards e-commerce sample data. The `sum` sub-aggregation calculates the sum of all taxed revenue for each week. The `moving_fn` script then returns the greater of the two values previous to the current value, or NaN if two values are not available:

Reviewer: ```suggestion
The following example creates a date histogram with a one-week interval from the OpenSearch Dashboards e-commerce sample data. The `sum` sub-aggregation calculates the sum of all taxed revenue for each week. The `moving_fn` script then returns the greater of the two values previous to the current value or NaN if two values are not available:
```

> 

Reviewer: Line 312: Should `NaN` be in code font?

        },
        {
          "key_as_string": "2025-03-31T00:00:00.000Z",
          "key": 1743379200000,
          "doc_count": 1048,
          "the_sum": {
            "value": 79448.60546875
          },
          "the_movavg": {
            "value": null
          }
        },
        {
          "key_as_string": "2025-04-07T00:00:00.000Z",
          "key": 1743984000000,
          "doc_count": 1048,
          "the_sum": {
            "value": 78208.4296875
          },
          "the_movavg": {
            "value": 79448.60546875
          }
        },
        {
          "key_as_string": "2025-04-14T00:00:00.000Z",
          "key": 1744588800000,
          "doc_count": 1073,
          "the_sum": {
            "value": 81277.296875
          },
          "the_movavg": {
            "value": 79448.60546875
          }
        },
        {
          "key_as_string": "2025-04-21T00:00:00.000Z",
          "key": 1745193600000,
          "doc_count": 924,
          "the_sum": {
            "value": 70494.2578125
          },
          "the_movavg": {
            "value": 81277.296875
          }
        }
      ]
    }
  }
}
```
</details>

## Example: Moving average

> The `moving_fn` aggregation replaces the deprecated `moving_avg` aggregation. The `moving_fn` aggregation is similar to the `moving_avg` aggregation but is more versatile since it computes arbitrary functions, not only averages. All of the predefined `moving_avg` functions are implemented in `moving_fn` as well. 

Reviewer: ```suggestion
The `moving_fn` aggregation replaces the deprecated `moving_avg` aggregation. The `moving_fn` aggregation is similar to the `moving_avg` aggregation but is more versatile since it computes arbitrary functions instead of only averages. All of the predefined `moving_avg` functions are implemented in `moving_fn` as well. 
```
